Addition of the CPU Activity Agent to steer CPU frequency based upon CPU Compute Activity

lhlawson opened the pull request:
- Relates to #2363 feature request from github issues
- Fixes #2363 request from github issues.
- Requires #1977
- Resolves #2639

This PR introduces and agent that modulates the CPU frequency within a set of user specified bounds based upon the MSR::SCALABILITY_RATIO signal AND modulate the CPU Uncore frequency within a set of user specified bounds based upon the Memory Bandwidth Monitor feature.  

The user may also provide an optional input 'phi' which can bias the agent frequency selection towards energy efficiency (phi of 1.0) or performance (phi of 0.0).  The default behavior is phi = 0.5 for the full range of the agent.  

Additional Tasks for this PR: 
- [x] Integration tests based upon the run_cpu_activity_*.py files
- [x] Policy generation script to find the Fe (frequency efficient) policy input based on AIB (arithmetic intensity benchmark) frequency sweeps
- [x] Update documentation of simplified policy generation script
- [x] Policy generation script to characterize relevant parts of the platform based upon AIB frequency sweeps
- [x] Updates to the cpu_activity.py experiment script to allow user specified Fe and Fmax values for phi sweep (current default is frequency min to frequency max range)
- [x] Updates to the cpu_activity.py experiment script to allow user specified system characterization for phi sweep (current default is no characterization, which disables uncore frequency control) 
- [x] Update to GEOPM 2.0 RC alias changes
- [x] Phi frequency response analysis
- [x] Agent documentation

lhlawson committed:
Added cpu-ca as a tutorial/plugin

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

Added scalability ratio call, removed acnt & pcnt direct calls

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson committed:
Moved the CPUActivityAgent to the base src dir from tutorials

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

Added basic CPU Activity agent experiment and run scripts

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

Added code (unused) for 3 different approaches to mitigating network comm based perf deg

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

Added unconstrained frequency range to cpu_activity.py sweeps

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson committed:
Added first pass CPU-CA policy recommendation script

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson committed:
Added memory bandwidth metric characterization

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson committed:
Added MBM parsing and policy output

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson committed:
Refactored cpu activity agent policy recommendation script

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

renamed cpu activity agent policy recommendation script

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

Updated comments on cpu activity policy recommendation

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson committed:
Update to CPU Activity Agent to provide 2 sample network fix infrastructure, per core controls, and gen script defaults

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson committed:
Added Unit tests, First pass of generic policy handling based on frequency map agent approach

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson committed:
Updated cpu_activity policies

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson committed:
Fixed default map sizing and uncharacterized system behavior

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson committed:
Added workload experiment scripts to extra_dist

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson committed:
Added CPUActivity Agent documentation, updated report header to match

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson committed:
Added cpu_activity command line args

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson committed:
Apply phi to policy values

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson committed:
Simplified cpu-ca policy recommender by removing tolerance & perf deg analysis

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson committed:
Updated CPU-CA documentation

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,540 @@
+                qm_max_itr = std::prev(qm_max_itr, 1);
+            }
+
+            double qm_normalized = (double) m_qm_rate.at(domain_idx).signal /
+                                            qm_max_itr->second;
+
+            // This is intended to handle dived by zero, either numerator or
+            // denominator being NAN, and the uncharacterized case
+            if (std::isnan(qm_normalized) ||
+                m_qm_max_rate.size() == 0) {
+                qm_normalized = 1.0;
+            }
```
```suggestion
            double qm_normalized = 1
            
            // Handle divided by zero, either numerator or
            // denominator being NAN, and the un-characterized case
            if (!std::isnan(m_qm_rate.at(domain_idx).signal) &&
                !std::isnan(qm_max_itr->second) && 
                qm_max_itr->second != 0 &&         
                m_qm_max_rate.size() != 0) {
                qm_normalized =(double) m_qm_rate.at(domain_idx).signal /
                                            qm_max_itr->second;
            }
```

lhlawson committed:
Updated qm ctr error handling to catch divide by zero case

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson committed:
Updated cpu activity policy recommendation script comments & hardcoded region

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson commented on test/CPUActivityAgentTest.cpp:
```diff
@@ -0,0 +1,665 @@
+//    EXPECT_CALL(*m_platform_io, adjust(GPU_FREQUENCY_CONTROL_IDX, expected_freq)).Times(M_NUM_GPU);
+//    m_agent->adjust_platform(policy);
+//    //Check a frequency decision resulted in write batch being true
+//    EXPECT_TRUE(m_agent->do_write_batch());
```
```suggestion

```

lhlawson committed:
Cleanup of old commented code

lhlawson committed:
Added enforce_policy to CPU Activity Agent

Signed-off-by: lhlawson <lowren.h.lawson@intel.com>

lhlawson committed:
Removed sample period from CPU Activity Agent

Signed-off-by: lhlawson <lowren.h.lawson@intel.com>

lhlawson committed:
added phi selection to cpu_activity.py to shorten integration test time if required

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson committed:
Added cpu activity agent characterization integration test

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson committed:
Updated CPU Activity Agent signals and testing

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson committed:
Updated mbm policy handling and policy size

Signed-off-by: lhlawson <lowren.h.lawson@intel.com>

lhlawson committed:
Added gen_cpu_activity_policy_recommendation.py to uncore_frequency_sweep

Signed-off-by: lhlawson <lowren.h.lawson@intel.com>

lhlawson committed:
Added check for workload executable

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson committed:
Update CPU CA signal names

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson committed:
Signal updates to high level aliases, coarser sweep in cpu activity agent integration test to shorten runtime

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson commented on src/CPUActivityAgent.hpp:
```diff
@@ -0,0 +1,116 @@
+
+            struct signal
+            {
+                int batch_idx;
+                double signal;
+                double sample;
```
```suggestion
                int batch_idx;
                double value;
```

lhlawson committed:
Removed unnecessary intermediate variable for initial agent

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson committed:
Signal update based on RC3

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson committed:
Cleaned up print statements in util.py

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson committed:
Integration test update for RC3 signals

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

bgeltz commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,532 @@
+                                              domain_idx), -1});
+        }
+
+        //Configuration of QM_CTR must match QM_CTR config used for training data
```
```suggestion
        // Configuration of QM_CTR must match QM_CTR config used for training data
```
Also this comment should explain what is being setup here.  For instance, what is EVENT_ID 2 and why do we care?

bgeltz commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,532 @@
+        ///////////////////////
+        //CPU POLICY CHECKING//
+        ///////////////////////
+        double freq_core_min = m_platform_io.read_signal("CPU_FREQUENCY_MIN_AVAIL", GEOPM_DOMAIN_BOARD, 0);
+        double freq_core_max = m_platform_io.read_signal("CPU_FREQUENCY_MAX_AVAIL", GEOPM_DOMAIN_BOARD, 0);
```
```validate_policy()``` gets called every time the Controller walks down.  Having these individual ```read_signal()``` calls in here will slow this down.  Do you need to read them every time through, or just once?

If it's just once, I'd put these values in a member variable, and cache them in ``init()`` like you do for the controls.

bgeltz commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,532 @@
+            in_policy[M_POLICY_UNCORE_FREQ_EFFICIENT] = m_freq_uncore_min;
+        }
+
+        if (in_policy[M_POLICY_UNCORE_FREQ_EFFICIENT] > in_policy[M_POLICY_UNCORE_FREQ_MAX]) {
```
```suggestion
        // When the policy is all NaNs, this check also verifies that
        // the system was not left in a bad state with regard to the
        // UNCORE_FREQUENCY_<MAX/MIN>_CONTROLs.
        if (in_policy[M_POLICY_UNCORE_FREQ_EFFICIENT] > in_policy[M_POLICY_UNCORE_FREQ_MAX]) {
```

bgeltz commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,532 @@
+             it != in_policy.end() && std::next(it) != in_policy.end(); std::advance(it, 2)) {
+            auto mapped_mem_bw = *(it + 1);
+            if (!std::isnan(*it)) {
+                // We are using a static cast rather than reinterpreting the
+                // memory so that regions can be input to this policy in the
+                // same form they are output from a report.
```
An example policy somewhere in the documentation would help explain this comment.  A one line example for the uncore freq policy field right here would also help.

bgeltz commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,532 @@
+    {
+        m_do_write_batch = false;
+
+        if (m_update_qm_max_rate) {
```
The body of this if() is only evaluated on the first call to ```adjust_platform()```.  ```m_update_qm_max_rate``` is set to true in the initializer list, and false at the end of this if().  Was this intentional?

lhlawson replied:
This was intentional.  The system characterization used by the agent in this version is based upon a single system wide characterization provided via policy and could be handled in the validate_policy section, but future updates in PR #2437 allow it to use per node signals (added in PR #2435) which necessitated moving the logic into adjust_platform to read the per node signals.

bgeltz commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,532 @@
+        m_resolved_f_uncore_max = in_policy[M_POLICY_UNCORE_FREQ_MAX];
+        double f_uncore_range = in_policy[M_POLICY_UNCORE_FREQ_MAX] - in_policy[M_POLICY_UNCORE_FREQ_EFFICIENT];
+
+        //m_resolved_f_uncore_efficient = 0;
```
```suggestion
```

bgeltz commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,532 @@
+            /////////////////////////////////////////////
+            // L3 Total External Bandwidth Measurement //
+            /////////////////////////////////////////////
+            auto qm_max_itr = m_qm_max_rate.lower_bound(uncore_freq);
+            if(qm_max_itr != m_qm_max_rate.begin()) {
+                qm_max_itr = std::prev(qm_max_itr, 1);
+            }
```
Please comment on what's going on here.

You're assign ```qm_max_itr``` an iterator that 1 *before* the lower bound of the current uncore frequency?  Why?

lhlawson replied:
This is intended to handle cases where the signal does not cleanly map to entries in the qm_max_rate keys (or vice versa).  
For a map of {1.9: a, 2.0: b, 2.2: c, 2.3: d} lower_bound(2.2) & lower_bound(2.1) will both return an iterator to (2.2: c).  Similarly lower_bound(2.21) will return an iterator to (2.3: d).  
As a simplification to address this issue while targeting energy efficiency we map to one iterator below the lower bound.

Alternative solutions or updates that may be implemented
- Only use std::prev if the uncore_freq doesn't match a key in the map
- Round (or floor) uncore_freq to a key in the m_qm_max_rate map before doing lower_bound
- Require map entries be valid values for the hardware specified by checking them vs the step size (not currently possible with existing GEOPM signals)

lhlawson replied:
"Only use std::prev if the uncore_freq doesn't match a key in the map" has been implemented in the latest set of changes.  Performance testing against 4 workloads of interest has shown no harm or better energy savings & perf deg for each workload.

bgeltz commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,532 @@
+            // For now only L3 bandwith metric is used.
+            double uncore_req = m_resolved_f_uncore_efficient + f_uncore_range * scalability_uncore;
+
+            //Clip uncore request within policy limits
```
Debug warn when this is happening?

lhlawson replied:
Moved to a report header following the discussion on the (similar) GPU compute activity agent clipping tracking [here](https://github.com/geopm/geopm/pull/2045#discussion_r871859791) as part of #2045

bgeltz commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,532 @@
+            }
+
+            double core_req = m_resolved_f_core_efficient + f_core_range * scalability;
+            //Clip core request within policy limits
```
Debug warn?

lhlawson replied:
Moved to a report header following the discussion on the (similar) GPU compute activity agent clipping tracking [here](https://github.com/geopm/geopm/pull/2045#discussion_r871859791) as part of #2045

bgeltz commented on integration/test/test_cpu_activity_agent.py:
```diff
@@ -0,0 +1,321 @@
+        """
+        Setup applications, execute, and set up class variables.
+        """
+
```
Add a comment explaining the setup here.

bgeltz commented on integration/test/test_cpu_activity_agent.py:
```diff
@@ -0,0 +1,321 @@
+@util.skip_unless_workload_exists("apps/minife/miniFE_openmp-2.0-rc3/src/miniFE.x")
+class TestIntegration_cpu_activity(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
```
This needs a refactor.

lhlawson committed:
Apply wsp & trimming dead code from review

Co-authored-by: Brad Geltz <brgeltz@gmail.com>

lhlawson committed:
Addressed review feedback on lower_bound, clipping, and variable inits

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson committed:
Added example policy to CPU Activity documentation

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson committed:
Fix CPUActivity Tests for lower_bound, added new lower_bound test, comment updates

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson committed:
Updated CPU activity agent report header

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

bgeltz commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,567 @@
+        m_freq_uncore_max = m_platform_io.read_signal("CPU_UNCORE_FREQUENCY_MAX_CONTROL", GEOPM_DOMAIN_BOARD, 0);
+        m_freq_core_min = m_platform_io.read_signal("CPU_FREQUENCY_MIN_AVAIL", GEOPM_DOMAIN_BOARD, 0);
+        m_freq_core_max = m_platform_io.read_signal("CPU_FREQUENCY_MAX_AVAIL", GEOPM_DOMAIN_BOARD, 0);
+
```
```suggestion
```

bgeltz commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,567 @@
+                                                                      domain_idx), NAN});
+        }
+
+        for (int domain_idx = 0; domain_idx < M_NUM_PACKAGE; ++domain_idx) {
```
```suggestion
```

bgeltz commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,567 @@
+            if (!std::isnan(*it)) {
+                // We are using a static cast rather than reinterpreting the
+                // memory so that regions can be input to this policy in the
+                // same form they are output from a report.
```
```suggestion
```

bgeltz commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,567 @@
+                // ...
+                // CPU_UNCORE_FREQ_<#>": 2400000000.0,
+                // "MAX_MEMORY_BANDWIDTH_<#>": 106613110000.0
+                auto uncore_freq = static_cast<uint64_t>(*it);
```
```suggestion
```

bgeltz commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,567 @@
+        std::set<double> policy_uncore_freqs;
+        for (auto it = in_policy.begin() + M_POLICY_FIRST_UNCORE_FREQ;
+             it != in_policy.end() && std::next(it) != in_policy.end(); std::advance(it, 2)) {
+            auto mapped_mem_bw = *(it + 1);
```
```suggestion
            auto uncore_freq = *it;
            auto mapped_mem_bw = *(it + 1);
```

bgeltz commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,567 @@
+
+            }
+        }
+
```
```suggestion
```

bgeltz commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,567 @@
+    {
+        m_do_write_batch = false;
+
+        if (m_update_qm_max_rate) {
```
```suggestion
        if (m_do_qm_max_rate_update) {
```

bgeltz commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,567 @@
+        m_resolved_f_uncore_efficient = 0;
+        m_resolved_f_uncore_max = 0;
+        m_resolved_f_core_efficient = 0;
+        m_resolved_f_core_max = 0;
```
Move this all to the initializer list in the constructor.

bgeltz commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,567 @@
+        m_resolved_f_core_efficient = 0;
+        m_resolved_f_core_max = 0;
+
+        // These are not currently guaranteed to be the system uncore min and max,
+        // just what the user/admin has previously set.
+        m_freq_uncore_min = m_platform_io.read_signal("CPU_UNCORE_FREQUENCY_MIN_CONTROL", GEOPM_DOMAIN_BOARD, 0);
+        m_freq_uncore_max = m_platform_io.read_signal("CPU_UNCORE_FREQUENCY_MAX_CONTROL", GEOPM_DOMAIN_BOARD, 0);
+        m_freq_core_min = m_platform_io.read_signal("CPU_FREQUENCY_MIN_AVAIL", GEOPM_DOMAIN_BOARD, 0);
+        m_freq_core_max = m_platform_io.read_signal("CPU_FREQUENCY_MAX_AVAIL", GEOPM_DOMAIN_BOARD, 0);
```
Move this all to init_platform_io.

bgeltz commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,567 @@
+        m_freq_core_max = m_platform_io.read_signal("CPU_FREQUENCY_MAX_AVAIL", GEOPM_DOMAIN_BOARD, 0);
+
+
+        init_platform_io();
```
Wrap this in a if level == 0 check.

bgeltz commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,567 @@
+                    // Not valid to have NAN max mem bw for uncore freq.
+                    GEOPM_DEBUG_ASSERT(!std::isnan(max_mem_bw),
+                                       "mapped CPU_UNCORE_FREQUENCY with no max memory bandwidth assigned.");
+                    m_qm_max_rate[uncore_freq] = max_mem_bw;
```
This seems like it can be moved into validate_policy(), then the if() block on 298 down to 320 can be removed from here.

bgeltz commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,567 @@
+            // For now only L3 bandwith metric is used.
+            double uncore_req = m_resolved_f_uncore_efficient + f_uncore_range * scalability_uncore;
+
+            //Clip uncore request within policy limits
```
```suggestion
            // Clip uncore request within policy limits
```

bgeltz commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,567 @@
+                m_uncore_freq_min_control.at(domain_idx).last_setting ||
+                uncore_freq_request.at(domain_idx) !=
+                m_uncore_freq_max_control.at(domain_idx).last_setting) {
+                //Adjust
```
```suggestion
                // Adjust
```

bgeltz commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,567 @@
+                m_platform_io.adjust(m_uncore_freq_max_control.at(domain_idx).batch_idx,
+                                    uncore_freq_request.at(domain_idx));
+
+                //save the value for future comparison
```
```suggestion
                // save the value for future comparison
```

bgeltz commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,567 @@
+            if (m_qm_max_rate.size() == 0) {
+                std::cerr << "Warning: <geopm> CPUActivityAgent did not receive a policy containing memory " <<
+                             "bandwidth characterization.  This may negatively impact agent performance." << std::endl;
+            }
```
Make this an error, then use the .size() to determine whether or not we need to build the map.

bgeltz committed:
Refactor test_cpu_activity_agent.py

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

lhlawson commented:
A follow on PR that refactorsthe performance model used in this agent into its own helper class (ActivityPerformanceModel.cpp) is underway on a fork [here](https://github.com/lhlawson/geopm/tree/public-lhlawson-node-ca) for the second version of the agent.  
The helper class is [here](https://github.com/lhlawson/geopm/blob/public-lhlawson-node-ca/src/ActivityPerformanceModel.cpp), and an agent using it [here](https://github.com/lhlawson/geopm/blob/public-lhlawson-node-ca/src/NodeActivityAgent.cpp#L474). 

This refactor does not require any user visible changes (such as modification to the policy) but has not been fully tested against a wide range of workloads at this time.

lhlawson committed:
Addressed review feedback

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson committed:
Integration test fixup

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson committed:
Updated unit test for new throw condition

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson commented on test/CPUActivityAgentTest.cpp:
```diff
@@ -0,0 +1,623 @@
+
+    //Adjust
+    //Check frequency
+    GEOPM_EXPECT_THROW_MESSAGE(m_agent->adjust_platform(policy), GEOPM_ERROR_INVALID,
```
```suggestion
    GEOPM_EXPECT_THROW_MESSAGE(m_agent->adjust_platform(policy), GEOPM_ERROR_LOGIC,
```

lhlawson committed:
Moved min/max freq reads to init for non-leaf agent policy checking

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson committed:
Addressed offline review comments on documentation

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

avilcheslopez commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,560 @@
+            in_policy[M_POLICY_CPU_FREQ_EFFICIENT] = m_freq_core_min;
+        }
+
+        if (in_policy[M_POLICY_CPU_FREQ_EFFICIENT] > m_freq_core_max ||
```
Very minor note: it seems this check is a bit redundant with the one below. In other words, it seems like the "if" below could be merged with this "if" taking advantage of the fact that a prior check ensures in_policy[M_POLICY_CPU_FREQ_MAX] is <= m_freq_core_max.

avilcheslopez commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,560 @@
+    {
+        m_do_write_batch = false;
+
+        if (m_qm_max_rate.size() == 0) {
```
Minor suggestion: ``m_qm_max_rate.empty()``

Same suggestion applies below.

avilcheslopez commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,560 @@
+    // Read signals from the platform and calculate samples to be sent up
+    void CPUActivityAgent::sample_platform(std::vector<double> &out_sample)
+    {
+        assert(out_sample.size() == M_NUM_SAMPLE);
```
GEOPM_DEBUG_ASSERT

lhlawson commented on service/docs/source/geopm_agent_cpu_activity.7.rst:
```diff
@@ -0,0 +1,238 @@
+options, where ``MAX_UNCORE_FREQ``, ``MIN_UNCORE_FREQ``,
+``MAX_CORE_FREQ``, and ``MIN_CORE_FREQ`` are replaced with the relevant
+system (or administrator chosen) values::
+
```
RMID values need to be set as well.

lhlawson committed:
Updates from offline review on gen policy & docs

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson committed:
fixed incorrect variable

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>

lhlawson commented on service/docs/source/geopm_agent_cpu_activity.7.rst:
```diff
@@ -0,0 +1,256 @@
+-----------------
+
+Generating a policy for the CPU Compute Activity agent requires system
+characterization using the GEOPM experiment infrastructure.  This can be
```
```suggestion
characterization using the GEOPM experiment infrastructure located in $GEOPM_SOURCE/integration/experiment/.  This can be
```

lhlawson commented on service/docs/source/geopm_agent_cpu_activity.7.rst:
```diff
@@ -0,0 +1,256 @@
+    gen_slurm.sh 1 arithmetic_intensity_benchmark uncore_frequency_sweep
+
+The generated test.sbatch should be modified to enable Memory Bandwidth
+Monitoring using the following::
```
```suggestion
Monitoring by adding the following to the test.sbatch file above the experiment script::
```

avilcheslopez commented on service/docs/source/geopm_agent_cpu_activity.7.rst:
```diff
@@ -0,0 +1,256 @@
+system (or administrator chosen) values::
+
+    --geopm-report-signals="MSR::QM_CTR_SCALED_RATE@package,CPU_UNCORE_FREQUENCY_STATUS@package,MSR::CPU_SCALABILITY_RATIO@package,CPU_FREQUENCY_MAX_CONTROL@package,CPU_UNCORE_FREQUENCY_MIN_CONTROL@package,CPU_UNCORE_FREQUENCY_MAX_CONTROL@package" \
+    --min-frequency=MIN_CORE_FREQ \
```
Suggestion (instead of using MIN_CORE_FREQ, ...): ``--min-frequency=<min system frequency>`` \

lhlawson commented on service/docs/source/geopm_agent_cpu_activity.7.rst:
```diff
@@ -0,0 +1,256 @@
+    --min-frequency=MIN_CORE_FREQ \
+    --max-frequency=MAX_CORE_FREQ \
+    --step-frequency=100000000 \
+    --max-uncore-frequency=MAX_UNCORE_FREQ \
+    --min-uncore-frequency=MIN_UNCORE_FREQ \
```
```suggestion
    --min-uncore-frequency=MIN_UNCORE_FREQ \
    --max-uncore-frequency=MAX_UNCORE_FREQ \
```

lhlawson commented on service/docs/source/geopm_agent_cpu_activity.7.rst:
```diff
@@ -0,0 +1,256 @@
+    --min-uncore-frequency=MIN_UNCORE_FREQ \
+    --step-uncore-frequency=100000000 \
+    --trial-count=5 \
+
```
Add note about increasing runtime in SBATCH file to an appropriately large value

lhlawson commented on service/docs/source/geopm_agent_cpu_activity.7.rst:
```diff
@@ -0,0 +1,256 @@
+    --step-uncore-frequency=100000000 \
+    --trial-count=5 \
+
+Then the test.sbatch script should be run on the node of interest.  This
```
```suggestion
Then the test.sbatch script should be run on the node of interest.  Using::

sbatch -w <NODE OF INTEREST> test.sbatch

This
```

lhlawson commented on service/docs/source/geopm_agent_cpu_activity.7.rst:
```diff
@@ -0,0 +1,256 @@
+characterization using the GEOPM experiment infrastructure.  This can be
+accomplished by running::
+
+    gen_slurm.sh 1 arithmetic_intensity_benchmark uncore_frequency_sweep
```
```suggestion
    gen_slurm.sh 1 arithmetic_intensity uncore_frequency_sweep
```

avilcheslopez replied:
NOTE: need to mention that AIB needs to be built

lhlawson commented on service/docs/source/geopm_agent_cpu_activity.7.rst:
```diff
@@ -0,0 +1,256 @@
+will run multiple kernels of varying intensity that stress the core and
+uncore to help with system characterization.
+
+The CPU compute activity agent policy can then be generated by running::
```
```suggestion
After sourcing the $GEOPM_SOURCE/integration/config/run_env.sh file
the CPU compute activity agent policy can then be generated by running::
```

avilcheslopez commented on service/docs/source/geopm_agent_cpu_activity.7.rst:
```diff
@@ -0,0 +1,256 @@
+Description
+-----------
+
+This agent is an experimental agent that is expected change significantly between
```
...that is expected ^to change...

avilcheslopez commented on service/docs/source/geopm_agent_cpu_activity.7.rst:
```diff
@@ -0,0 +1,256 @@
+
+This agent is an experimental agent that is expected change significantly between
+GEOPM 2.0 and 2.1.  Some examples of changes that may occur include agent policy
+changes, algorithmic chagnes, and folding the agent into other agents that achieve
```
typo: chagnes

avilcheslopez commented on service/docs/source/geopm_agent_cpu_activity.7.rst:
```diff
@@ -0,0 +1,256 @@
+The ``CPU_UNCORE_ACTIVITY`` is defined as a simple ratio of the current memory bandwdith
+divided by the maximum possible bandwidth at the current ``CPU_UNCORE_FREQUENCY_STATUS`` value.
+
+The ``Fe`` value for all domains isintended to be an energy efficient frequency
```
typo: isintended

avilcheslopez commented on service/docs/source/geopm_agent_cpu_activity.7.rst:
```diff
@@ -0,0 +1,256 @@
+energy consumption for the workload.
+
+``Fmax`` is intended to be the maximum allowable frequency for the domain,
+and may be set as the domain's default  maximum frequency, or limited based
```
Extra space between "default" and "maximum".

avilcheslopez commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,560 @@
+
+        // Check for NAN to set default values for policy
+        if (std::isnan(in_policy[M_POLICY_CPU_FREQ_EFFICIENT])) {
+            in_policy[M_POLICY_CPU_FREQ_EFFICIENT] = m_freq_core_min;
```
Q: why not take the midpoint between the min and the max as the default?

lhlawson replied:
That approach may also work but was found in testing to be less beneficial than either:
1) Sticker - 2*freq step
2) Min freq

avilcheslopez commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,560 @@
+            in_policy[M_POLICY_UNCORE_FREQ_MAX] = m_freq_uncore_max;
+        }
+        if (std::isnan(in_policy[M_POLICY_UNCORE_FREQ_EFFICIENT])) {
+            in_policy[M_POLICY_UNCORE_FREQ_EFFICIENT] = m_freq_uncore_min;
```
Same question as before

avilcheslopez commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,560 @@
+    // Indicate whether to send the policy down to children
+    bool CPUActivityAgent::do_send_policy(void) const
+    {
+        return true;
```
Q: always return true or only send policy to children once?

avilcheslopez commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,560 @@
+            // and the uncore_freq is not an exact match.
+            if(qm_max_itr != m_qm_max_rate.begin() &&
+               m_qm_max_rate.find(uncore_freq) == m_qm_max_rate.end()) {
+                qm_max_itr = std::prev(qm_max_itr, 1);
```
qm_max_itr--?

avilcheslopez committed:
Make the CPU Activity Agent a beta feature

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Improve CPUActivityAgent documentation, make minor fixes in code

- Improvements to documentation to make system characterization steps
  clearer
- Minor code clean-up and fixes

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Minor clean-up in CPUActivityAgent unit tests

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Fix build issue in CPUActivityAgentTest

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

dannosliwcd commented on Makefile.am:
```diff
@@ -457,7 +457,9 @@ libgeopmpolicy_la_SOURCES = src/Accumulator.cpp \
                             # end
 
-beta_source_files = src/Daemon.cpp \
+beta_source_files = src/CPUActivityAgent.cpp \
```
We should probably add a [.. note::](https://sublime-and-sphinx-guide.readthedocs.io/en/latest/notes_warnings.html) block to the .rst files for these files that specifies they are behind the beta flag.

avilcheslopez replied:
Yep! Will do.

dannosliwcd commented on integration/experiment/energy_efficiency/cpu_activity.py:
```diff
@@ -0,0 +1,110 @@
+    parser.add_argument('--cpu-frequency-efficient', dest='cpu_fe',
+                        action='store', default='nan',
+                        help='The efficient frequency of the cpu (a.k.a. Fce) for this experiment')
+    parser.add_argument('--cpu-frequency-max', dest='cpu_fmax',
```
May be confusing vs `--max-frequency` in `common_args.py` (which is a max used in a sweep range). Maybe "--cpu-ca-max-frequency'? Maybe it doesn't matter.

avilcheslopez replied:
You know what... Looking at ``common_args.py``, I'm thinking it deserves a second look and perhaps some refactoring is in order (after renaming some of the arguments). At the moment, the meaning of "frequency" in ``common_args.py`` is dependent on the experiment script that uses it (could be core or uncore, for example).

That being the case, I'm tempted to leave these (``cpu_activity``) arguments as they are because this agent needs both: a core frequency limit and an uncore frequency limit. Then we can revisit this once we take a second look at ``common_args.py``.

avilcheslopez replied:
Filed an issue to look into this: #2648

dannosliwcd commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,564 @@
+                                                             domain_idx), NAN});
+            m_core_freq_control.push_back({m_platform_io.push_control("CPU_FREQUENCY_MAX_CONTROL",
+                                                                      GEOPM_DOMAIN_CORE,
+                                                                      domain_idx), -1});
```
Why -1 instead of NaN?

avilcheslopez replied:
In ``CPUActivityAgent::adjust_platform()``, when it determines what the frequency should be (``core_request``), it needs to compare it against the last setting to decide whether the setting actually needs to be updated. Using -1 allows that check to work properly the first time a frequency decision is made.

dannosliwcd replied:
I think `NaN` would be a more defensive way to meet that goal.  This placeholder value is not expected to be written. If we use NaN instead, then `PlatformIOImp::adjust` will throw an exception if that ever does happen, instead of silently writing the placeholder value.

With the no-fast-math change that just went in, we can assume `NaN != <anything>`, so the `CPUActivityAgent::adjust_platform` check should still work with NaN.

dannosliwcd replied:
Actually, I wonder if this should be delegated to `FrequencyGovernor` instead.

avilcheslopez committed:
Mention CPUActivityAgent is a beta feature in docs

Also, fix incorrect references to other documents in CPUActivityAgent
docs after the rebase.

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

dannosliwcd commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,564 @@
+            if (m_qm_max_rate.empty()) {
+                throw Exception("CPUActivityAgent::" + std::string(__func__) +
+                                "(): CPUActivityAgent policy did not contain" +
+                                " memory bandwidth characteriztaion.",
```
```suggestion
                                " memory bandwidth characterization.",
```

dannosliwcd commented on src/CPUActivityAgent.hpp:
```diff
@@ -0,0 +1,118 @@
+            };
+
+            std::map<std::string, double> m_policy_available;
+            std::map<double,double> m_qm_max_rate;
```
Maybe add a comment like: This maps (uncore frequency) -> (maximum memory bandwidth) specified in the input policy

dannosliwcd commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,564 @@
+            if(qm_max_itr != m_qm_max_rate.begin() &&
+               m_qm_max_rate.find(uncore_freq) == m_qm_max_rate.end()) {
+                qm_max_itr--;
+            }
```
I think this means we want the last element <= uncore_freq.

Could probably simplify with something like:
```
qm_max_itr = upper_bound(uncore_freq)
if (not the beginning) --qm_max_itr;
```

dannosliwcd commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,564 @@
+        }
+
+        // Set per core controls
+        for (int domain_idx = 0; domain_idx < M_NUM_CORE; ++domain_idx) {
+            if(std::isnan(core_freq_request.at(domain_idx))) {
+                core_freq_request.at(domain_idx) = in_policy[M_POLICY_CPU_FREQ_MAX];
+            }
+            if (core_freq_request.at(domain_idx) !=
+                m_core_freq_control.at(domain_idx).last_setting) {
+                // Adjust
+                m_platform_io.adjust(m_core_freq_control.at(domain_idx).batch_idx,
+                                     core_freq_request.at(domain_idx));
+
+                // Save the value for future comparison
+                m_core_freq_control.at(domain_idx).last_setting = core_freq_request.at(domain_idx);
+                ++m_core_frequency_requests;
+                m_do_write_batch = true;
+            }
+        }
```
Bounds-clamping, sending to PIO, and flagging whether write_batch is needed could be delegated to `geopm::FrequencyGovernor`

dannosliwcd replied:
Though I guess that would only work if the governor runs on a system that natively uses core domain for frequency controls. Otherwise, all of the M_NUM_CORE loops here need to change to ask the governor which  `FrequencyGovernor::frequency_domain_type()` to use, or we need to tell the governor which domain we want (latter option doesn't currently exist, but is in one of my upcoming PRs for the balancer agent).

dannosliwcd replied:
From our offline discussion, let's defer this comment thread because it also depends on other work:

 - Optional: #2642 would enable you to use FrequencyGovernor without modifying your M_NUM_CORE loops to ask the governor what to use.
 - #2643 to resolve this issue, and to resolve the discovery that FrequencyMapAgent also does not use the helper

dannosliwcd commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,564 @@
+    void CPUActivityAgent::wait(void)
+    {
+        geopm_time_s current_time;
+        do {
+            geopm_time(&current_time);
+        }
```
Do we really need a busy wait? It seems like this might be done to reduce variation in time between controller steps. Does this agent actually need that? All of the sampled values are read as rates anyways (vs point-in-time snapshots of monotonic counters).

avilcheslopez replied:
Yeah, this is just the standard wait implementation all agents are using. We've talked about standardizing the implementation and factoring it out so that it doesn't have to be implemented in every agent. We've talked about a smarter implementation that leverages nanosleep() to minimize the amount of time we rely on a busy wait.

I would prefer to defer addressing this to when this refactoring work is done, as it will address it for all the agents.

dannosliwcd replied:
Sounds good to defer this comment to #1065

avilcheslopez committed:
Minor code clean-up in CPUActivityAgent

Made minor improvements in code, fixed a typo.

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

dannosliwcd commented on src/CPUActivityAgent.cpp:
```diff
@@ -324,19 +324,14 @@ namespace geopm
+            // is <= uncore_freq.
+            auto qm_max_itr = m_qm_max_rate.upper_bound(uncore_freq);
+            if (qm_max_itr != m_qm_max_rate.begin())
                 qm_max_itr--;
```
```suggestion
            if (qm_max_itr != m_qm_max_rate.begin()) {
                --qm_max_itr;
            }
```

or on one line: `if (qm_max_itr != m_qm_max_rate.begin()) --qm_max_itr;`

Not a required change, since this feels like a style preference issue and I don't know if the team has any rules about this. My personal motivation is that I've seen bugs appear on projects from moving code into or out the body of this style of if-statement during later changes.

avilcheslopez replied:
Yeah, I don't actually need post-increment semantics here (was mainly following common style). In the rest of the codebase, for similar operations, we have a bit of a mix, but in the vast majority of cases we use pre-increment. I'll update this accordingly.

dannosliwcd commented on integration/experiment/uncore_frequency_sweep/gen_cpu_activity_policy_recommendation.py:
```diff
@@ -0,0 +1,169 @@
+    """
+    df_filtered = df
+    if region_list:
+        df_filtered = df[df['region'].isin(region_list)]
+
+    if ('MSR::QM_CTR_SCALED_RATE' not in df_filtered.columns):
+        df_filtered['MSR::QM_CTR_SCALED_RATE'] = (df_filtered['MSR::QM_CTR_SCALED_RATE@package-0'] +
+                                             df_filtered['MSR::QM_CTR_SCALED_RATE@package-1'])/2
```
Does this print a warning about setting a value in a copied slice (vs setting in a slice of the original).

If it does, then maybe modify the start of the selection to:
```suggestion
        df_filtered = df.loc[df['region'].isin(region_list)]

    if ('MSR::QM_CTR_SCALED_RATE' not in df_filtered.columns):
        df_filtered['MSR::QM_CTR_SCALED_RATE'] = (df_filtered['MSR::QM_CTR_SCALED_RATE@package-0'] +
                                             df_filtered['MSR::QM_CTR_SCALED_RATE@package-1'])/2
```

avilcheslopez replied:
Will fix!

dannosliwcd commented on integration/experiment/uncore_frequency_sweep/gen_cpu_activity_policy_recommendation.py:
```diff
@@ -0,0 +1,169 @@
+    mem_bw_dict = {}
+    for k in uncore_freq_set:
+        mem_df = df.groupby('uncore-frequency (Hz)').get_group(k)
+        #TODO: use both packages!  This will skew in favor of package-0
```
Is this still needed? Should we open an issue?

lhlawson replied:
This shouldn't be needed currently, but is a future feature request if we find there is a significant amount of package to package variability that we can capitalize on for better energy efficiency.

dannosliwcd commented on integration/experiment/uncore_frequency_sweep/gen_cpu_activity_policy_recommendation.py:
```diff
@@ -0,0 +1,169 @@
+        fixed_freq_col = 'uncore-frequency (Hz)'
+    else:
+        sys.stderr.write('<geopm> Error: unsupported domain ' + domain + \
+                         'proovided\n')
```
```suggestion
                         'provided\n')
```

avilcheslopez committed:
Minor code clean-up for CPU-CA agent, its experiment infrastructure

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Add comments in test_cpu_activity_agent setup

Add comments to clarify the integration test's global setup steps.

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

bgeltz commented on service/docs/source/GEOPM_CXX_MAN_CPUActivityAgent.3.rst:
```diff
@@ -0,0 +1,34 @@
+
```
```suggestion
```

bgeltz commented on service/docs/source/GEOPM_CXX_MAN_CPUActivityAgent.3.rst:
```diff
@@ -0,0 +1,34 @@
+
+geopm::CPUActivityAgent(3) -- agent for selecting CPU frequency based on CPU compute activity
+=============================================================================================
+
```
```suggestion
```

bgeltz commented on service/docs/source/geopm_agent_cpu_activity.7.rst:
```diff
@@ -0,0 +1,270 @@
+
```
```suggestion
```

bgeltz commented on service/docs/source/geopm_agent_cpu_activity.7.rst:
```diff
@@ -0,0 +1,270 @@
+
+
+
+
```
```suggestion
```

bgeltz commented on service/docs/source/geopm_agent_cpu_activity.7.rst:
```diff
@@ -0,0 +1,270 @@
+    algorithm. It is also possible that this agent may be refactored and
+    combined with other agents.
+
+The goal of this Agent is to save CPU energy by scaling CPU frequency based upon
```
Lowercase unless you're talking about the Agent class/interface.  This seems to be the pattern we use in other files AFIACT.

```suggestion
The goal of this agent is to save CPU energy by scaling CPU frequency based upon
```

Or you could spell out the full agent name as you do elseware.

avilcheslopez replied:
Opted to spell out the name of the agent in the beginning (first reference) and simply refer to it as "the agent" afterwards.

bgeltz commented on service/docs/source/geopm_agent_cpu_activity.7.rst:
```diff
@@ -0,0 +1,270 @@
+replaced with relevant system (or administrator chosen) values::
+
+    --geopm-report-signals="MSR::QM_CTR_SCALED_RATE@package,CPU_UNCORE_FREQUENCY_STATUS@package,MSR::CPU_SCALABILITY_RATIO@package,CPU_FREQUENCY_MAX_CONTROL@package,CPU_UNCORE_FREQUENCY_MIN_CONTROL@package,CPU_UNCORE_FREQUENCY_MAX_CONTROL@package" \
+    --min-frequency=<min. core frequency> \
```
```suggestion
    --min-frequency=<min core frequency> \
```

Either min or minimum.  Do this for all max as well.

bgeltz commented on service/docs/source/geopm_agent_cpu_activity.7.rst:
```diff
@@ -0,0 +1,270 @@
+    Before performing the system characterization, please ensure the
+    system is quiesced (i.e. not running other heavy processes/workloads).
+
+The first step is to generate the execution script by running::
```
```suggestion
The first step is to generate the execution script by running:
```

Does this need to be a double colon?

dannosliwcd replied:
The colons are not rendered, but the indented block following the double colon is treated as preformatted text. I think without the `::`, the indented text would be rendered as a block quote.

avilcheslopez replied:
I tested it out: without the double colon the following block of text appears as normal text (i.e. not preformatted text). At least it still appears indented.

Leaving the double colon.

bgeltz commented on service/docs/source/geopm_agent_cpu_activity.7.rst:
```diff
@@ -0,0 +1,270 @@
+
+Additionally the ``test.sbatch`` should be modified to include the following
+experiment options, where the text within angle brackets (``<>``) needs to be
+replaced with relevant system (or administrator chosen) values::
```
Include examples of how to derive this information with geopmread.

bgeltz commented on src/CPUActivityAgent.hpp:
```diff
@@ -0,0 +1,120 @@
+            std::map<std::string, double> m_policy_available;
+            // Maps uncore frequency -> maximum memory bandwidth achieved by
+            // that frequency (determined by system characterization)
+            std::map<double,double> m_qm_max_rate;
```
```suggestion
            std::map<double, double> m_qm_max_rate;
```

bgeltz commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,559 @@
+
+    // Distribute incoming policy to children
+    void CPUActivityAgent::split_policy(const std::vector<double>& in_policy,
+                                    std::vector<std::vector<double> >& out_policy)
```
```suggestion
    void CPUActivityAgent::split_policy(const std::vector<double>& in_policy,
                                        std::vector<std::vector<double> >& out_policy)
```

bgeltz commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,559 @@
+
+        // Set per core controls
+        for (int domain_idx = 0; domain_idx < M_NUM_CORE; ++domain_idx) {
+            if(std::isnan(core_freq_request.at(domain_idx))) {
```
```suggestion
            if (std::isnan(core_freq_request.at(domain_idx))) {
```

bgeltz commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,559 @@
+
+        // Set per package controls
+        for (int domain_idx = 0; domain_idx < M_NUM_PACKAGE; ++domain_idx) {
+            if(std::isnan(uncore_freq_request.at(domain_idx))) {
```
```suggestion
            if (std::isnan(uncore_freq_request.at(domain_idx))) {
```

bgeltz commented on src/CPUActivityAgent.cpp:
```diff
@@ -0,0 +1,559 @@
+        return false;
+    }
+
+    void CPUActivityAgent::adjust_platform(const std::vector<double>& in_policy)
```
This method is great candidate for a refactor.  It's pretty long, and can be broken up into helper functions.  Hopefully we can clean this up with the PerformanceModel work in the near future.

bgeltz commented on integration/test/test_cpu_activity_agent.py:
```diff
@@ -0,0 +1,269 @@
+        if sys.exc_info() != (None, None, None):
+            TestIntegration_cpu_activity._keep_files = True
+
+    @util.skip_unless_config_enable('beta')
```
This should be added up at the class level (and removed from here) to ensure that setUpClass() does not run if beta is not enabled.  Seems like it would take a long time given how many combinations of agents and runs there are.

bgeltz commented on test/CPUActivityAgentTest.cpp:
```diff
@@ -0,0 +1,583 @@
+    ASSERT_EQ(m_mbm_max.size(), M_NUM_UNCORE_MBM_READINGS);
+
+    for (size_t i = 0; i < M_NUM_UNCORE_MBM_READINGS; ++i) {
+        m_default_policy.push_back(static_cast<double>(m_cpu_uncore_freqs[i]));
```
Why is the static_cast necessary? m_cpu_uncore_freqs is already a vector of doubles.

bgeltz commented on test/CPUActivityAgentTest.cpp:
```diff
@@ -0,0 +1,583 @@
+    std::vector<double> policy;
+
+    // default policy with 1.2-2.4GHz MBM
+    // max rates defined is accepted
```
```suggestion
    // max rates defined are accepted
```

avilcheslopez committed:
CPUActivityAgent: code clean-up and improvements to documentation

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

bgeltz committed:
Spacing fixes

- Added missing config.h.
- Fix sorting.

Signed-off-by: Brad Geltz <brad.geltz@intel.com>
Change-Id: Ie8c62647f340dbeb440586f5582905c48be80332
